core/arch/arm/sm/sm_a32.S

```diff
@@ -312,6 +312,33 @@ UNWIND(	.fnstart)
 	cps	#CPSR_MODE_MON
 	/* Point just beyond sm_ctx.sec */
 	sub	sp, r0, #(SM_CTX_SIZE - SM_CTX_NSEC)
+
+#ifdef CFG_INIT_CNTVOFF
+	read_scr r0
+	orr	r0, r0, #SCR_NS /* Set NS bit in SCR */
+	write_scr r0
+
+	/*
+	 * Accessing CNTVOFF:
+	 * If the implementation includes the Virtualization Extensions
+	 * this is a RW register, accessible from Hyp mode, and
+	 * from Monitor mode when SCR.NS is set to 1.
+	 * If the implementation includes the Security Extensions
+	 * but not the Virtualization Extensions, an MCRR or MRRC to
+	 * the CNTVOFF encoding is UNPREDICTABLE if executed in Monitor
+	 * mode, regardless of the value of SCR.NS.
+	 */
+	read_idpfr1 r2
+	ands	r2, r2, #IDPFR1_GENTIMER_MASK
```

Comment:
Shouldn't you test also IDPFR1[virtualiation] ?
```
	read_idpfr1	r2
	ands		r2, r2, #IDPFR1_GENTIMER_MASK
+	andsne		r2, r2, #IDPFR1_VIRT_MASK
	beq		.no_gentimer
	mov		r2, #0
	write_cntvoff 	r2, r2

Reply:
use this `ands	  r2, r2, #(IDPFR1_GENTIMER_MASK | IDPFR1_VIRT_MASK)`?

Reply:
I think you need `(r2 & (gentimer_flag | virt_flag)) == (gentimer_flag | virt_flag)`.
Is my proposal wrong? It should do:
```c
	if (idpfr1[gentimer_flag] != 0 &&
	    idpfr1[virt_flag] != 0)
		write_cntvoff(0);
```